### PR TITLE
Modularize Dreambooth LoRA SD inferencing during and after training

### LINKED ISSUE #6545: [Tracker] modularize inferencing during and after training in the example scripts
We provide support for running validation inference during and after training in our officially maintained training examples. This is very helpful to keep track of the training progress. 

We could modularize some bits in the example to reduce the LoC. 

The `train_lcm_distill_lora_sdxl.py` [script](https://github.com/huggingface/diffusers/blob/main/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py) already does this:

https://github.com/huggingface/diffusers/blob/33d2b5b08764d2579dd4ea555821b8986cd2dcf9/examples/consistency_distillation/train_lcm_distill_lora_sdxl.py#L100

It would be nice to follow something similar for the rest of the scripts too. Here's a handy list of the scripts where we'd like to incorporate this change:

- DreamBooth
  - [x] [SD LoRA](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py)
  - [x] [SDXL LoRA](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py)
- InstructPix2Pix
  - [ ] [SD](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix.py)
  - [x] [SDXL](https://github.com/huggingface/diffusers/blob/main/examples/instruct_pix2pix/train_instruct_pix2pix_sdxl.py)
- Text-to-image
  - [ ] [SD](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py)
  - [x] [SD LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py)
  - [x] [SDXL LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora_sdxl.py)

Feel free to comment here if you're interested. 

Also, when opening PRs, please target one example at a time and please tag me in the PRs.

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Part of #6545 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yiyixuxu** [comment]
cc @sayakpaul 
should we merge this in?

**Review by sayakpaul — COMMENTED** [review]
Looks very good. Thanks a lot for working on this. 

Could you also provide some sample training commands with which we could test this?

**Comment by rootonchair** [comment]
Hi @sayakpaul, you could test it with the following line
````
accelerate launch examples/dreambooth/train_dreambooth_lora.py --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" --instance_data_dir="dog" --output_dir=dog_lora --instance_prompt="a photo of sks dog" --train_batch_size=1 --resolution=512 --checkpointing_steps=100 --learning_rate=1e-4 --max_train_steps=500 --validation_prompt="a photo of sks dog in a bucket" --validation_epochs=50 --seed="0"
````

File: examples/dreambooth/train_dreambooth_lora.py [hunk 1]
```diff
@@ -1255,12 +1322,9 @@ def compute_text_embeddings(prompt):
             if global_step >= args.max_train_steps:
                 break
 
+        images = []
```
[the remark this answers is not in the document]
Comment (rootonchair) [reply examples/dreambooth/train_dreambooth_lora.py:1333]: We don't. I will delete this line

File: examples/dreambooth/train_dreambooth_lora.py [hunk 2]
```diff
@@ -66,6 +66,9 @@
 from diffusers.utils.torch_utils import is_compiled_module
 
 
+if is_wandb_available():
+    import wandb
+
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:71]:
```suggestion
```

File: examples/dreambooth/train_dreambooth_lora.py [hunk 3]
```diff
@@ -684,7 +752,6 @@ def main(args):
     if args.report_to == "wandb":
         if not is_wandb_available():
             raise ImportError("Make sure to install wandb if you want to use it for logging during training.")
-        import wandb
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:687]:
```suggestion
        import wandb
```

**Review by sayakpaul — APPROVED** [review]
Just one change requested. Thanks a lot for working on this!

Ran with your command:

```bash
accelerate launch train_dreambooth_lora.py --pretrained_model_name_or_path="runwayml/stable-diffusion-v1-5" --instance_data_dir="dog" --output_dir=dog_lora --instance_prompt="a photo of sks dog" --train_batch_size=1 --resolution=512 --checkpointing_steps=100 --learning_rate=1e-4 --max_train_steps=500 --validation_prompt="a photo of sks dog in a bucket" --validation_epochs=50 --seed="0" --report_to="wandb"
```

Working perfectly: https://wandb.ai/sayakpaul/dreambooth-lora/runs/ligw3fg4

**Comment by sayakpaul** [comment]
Thanks a lot for working on this!

**Pull request merged**